### ISSUE TITLE
It looks like generated equals() method contains the code which is never executed.

### ISSUE THREAD

**Issue by xeagle2** [rzwitserloot/lombok#1192]
I use lombok 1.16.8 and when I use annotation to generate equals() it generates the following (trying to describe schematically): 

```
public boolean equals(Object o) {
    if (o == this) {
      return true;
    }
    if (!(o instanceof CLASS)) {
      return false;
    }
    CLASS other = (CLASS)o;
    if (!other.canEqual(this)) {
      return false;
    }
    ...
}

protected boolean canEqual(Object other) {
    return other instanceof CLASS;
  }
```

It looks like the following block is never executed

```
if (!other.canEqual(this)) {
      return false;
}
```
because the method terminates on the condition !(o instanceof CLASS).

What do you think?

**Comment by SOF3** [comment]
Consider this code.

```java
@EqualsAndHashCode
class Foo {
  public int a = 1;
}
@EqualsAndHashCode(callSuper = true) // whether callSuper is true actually doesn't really matter
class Bar extends Foo {
  public int b = 2;
}

Foo a = new Foo(), b = new Bar();
```

Without `canEqual`, `a.equals(b)` will call the `Foo.equals` method, which only checks the field `a`. OK, it returns true, because both have `a = 1`.
What about `b.equals(a)`? `b` will check if `a instanceof Bar`, which is false, so it will return false.

`a.equals(b)` returns true and `b.equals(a)` returns false. This is wrong, because `equals` must be symmetric.

**Comment by xeagle2** [comment]
Thanks for the quick and meaningful reply. I missed this for some reason. Just read your case and admit you are absolutely right.

Thanks again for getting the light on this!

**Comment by rodrigo797** [comment]
Hi, considering the above example.

@EqualsAndHashCode
class Foo {
  public int a = 1;
}
@EqualsAndHashCode(callSuper = true) // whether callSuper is true actually doesn't really matter
class Bar extends Foo {
  public int b = 2;
}

Foo a = new Foo(), b = new Bar();
a.equals(b); // true
b.equals(a); // false

How can I give coverage 100% of EqualsAndHashCode of Foo class?

Making test below.
    @Test
    public void canEquals() {
        Foo a = new Foo(), b = new Bar();
        assertFalse(a.equals(b));
        assertFalse(b.equals(a));
        assertFalse(a.equals(null));
        assertFalse(b.equals(null));
        a.setA(0);
        b.setA(0);
        assertFalse(b.equals(a));
        assertFalse(a.equals(b));
        assertTrue(a.equals(a));
        assertTrue(b.equals(b));
    }
![foo test](https://cloud.githubusercontent.com/assets/5619199/24424378/c023a22e-13d6-11e7-8997-e2dc4e0a7688.PNG)
![bar test](https://cloud.githubusercontent.com/assets/5619199/24424381/c48313f4-13d6-11e7-93e0-d4db2dc20f3c.PNG)

**Comment by obecker** [comment]
@rodrigo797 I guess you have to create two `Foo` instances with different values for `a` and invoke `foo1.equals(foo2)`.

**Comment by obecker** [comment]
I have a comment regarding the initial issue: In my scenario I have a class `A` and a *final* class `B extends A`, both have the `EqualsAndHashCode` annotation. While I understand that both `A` and `B` need the `canEqual` method, it is not clear why the generated `equals` method in `B` also invokes `other.canEqual(this)`. Since `B` is final, there exists no other object that is both instance of `B` and has a different `canEqual` method. So in my scenario I also have a coverage problem. Normally I handle these problems by adding another sub class just for testing purposes (which I cannot do here since the class is final ...).

**TL;DR**: if a final class has an `EqualsAndHashCode` annotation, the generated `equals` method should not invoke `other.canEqual(this)`
